backend_jobs/ingestion_update_visibility/pipeline_lib/firestore_database.py

```diff
@@ -0,0 +1,135 @@
+            A list of dictionaries with all the information (fields and id)
+            of each one of the Firestore query's image documents.
+        """
+        _valedate_one_arg(image_provider, pipeline_run)
+        # the lower limit for querying the database by the random field.
+        random_min = element * constance.RANGE_OF_BATCH
+        # the higher limit for querying the database by the random field.
+        random_max = random_min + constance.RANGE_OF_BATCH
+        if image_provider:
+            query = self.db.collection_group(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS)\
+                .where(
+                    database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_PROVIDER_ID,
+                    u'==',
+                    image_provider)\
+                .where(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_RANDOM, u'>=', random_min)\
+                .where(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_RANDOM, u'<', random_max)\
+                .stream()
+        else:
+            query = self.db.collection_group(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS)\
+                .where(
+                    database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_PIPELINE_RUN_ID,
+                    u'==',
+                    pipeline_run)\
+                .where(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_RANDOM, u'>=', random_min)\
+                .where(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_RANDOM, u'<', random_max)\
+                .stream()
+        return (_add_id_to_dict(doc) for doc in query)
+
+
+class UpdateVisibilityInDatabase(apache_beam.DoFn):
+    """ Updates Firestore Database visibility field to visible.
+    Updates visibility inside pipeline document in 'PipelineRuns' subcollection
+    and in the 'Images' collection.
+    """
+
+    def setup(self):
+        self.db = initialize_db()
+
+    def process(self, element, visibility):
+        """This function updates the visibility in the Images/ PipelineRun firestore database.
+
+        Args:
+            element: A dictionaries with all the information (fields and id) to update.
+            visibility: The visibility we are updating the doc to, e.g. 'VISIBLE'/ 'INVISIBLE'
+        """
+        parent_image_id = element[database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS_FIELD_PARENT_IMAGE_ID]
+        parent_image_ref = self.db.collection(database_schema.COLLECTION_IMAGES)\
+            .document(parent_image_id)
+        doc_ref = parent_image_ref.collection(database_schema.COLLECTION_IMAGES_SUBCOLLECTION_PIPELINE_RUNS)\
+            .document(element['id'])
+        parent_image_ref.update({
+            database_schema.COLLECTION_IMAGES_FIELD_VISIBILITY:
+                visibility.value  # TODO: fix functionality to be the max between all doc in subcollection.
```

Comment:
so this would overwrite the visibility even though another pipeline would have the image visible. Let's discuss what is the right behavior here

Reply:
fixed